### ISSUE TITLE
1.0.2r memory leak in GOST engine (with patch)

### ISSUE THREAD

**Issue by NMorozxov** [openssl/openssl#8559]
Memory leak in GOST Engine 
engines/gost/gost_ameth.c
in 
static int pub_decode_gost94(EVP_PKEY *pk, X509_PUBKEY *pub)
and
static int pub_encode_gost94(X509_PUBKEY *pub, const EVP_PKEY *pk)
if databuf malloc failed, the octec is not freed.
patch that fix this issue
```patch
diff -uNr openssl-1.0.2r.orig/engines/ccgost/gost_ameth.c openssl-1.0.2r/engines/ccgost/gost_ameth.c
--- openssl-1.0.2r.orig/engines/ccgost/gost_ameth.c	2019-03-20 14:29:50.239495797 +0300
+++ openssl-1.0.2r/engines/ccgost/gost_ameth.c	2019-03-22 13:34:41.668782005 +0300
@@ -618,6 +618,7 @@
     }
     databuf = OPENSSL_malloc(octet->length);
     if (databuf == NULL) {
+        ASN1_OCTET_STRING_free(octet);
         GOSTerr(GOST_F_PUB_DECODE_GOST94, ERR_R_MALLOC_FAILURE);
         return 0;
     }
@@ -693,6 +694,7 @@
     }
     databuf = OPENSSL_malloc(octet->length);
     if (databuf == NULL) {
+        ASN1_OCTET_STRING_free(octet);
         GOSTerr(GOST_F_PUB_DECODE_GOST01, ERR_R_MALLOC_FAILURE);
         return 0;
     }
```

**Comment by mspncp** [comment]
Thanks @NMorozxov for reporting this issue and providing a patch.

After taking a quick look at the sourcecode it seems to me that your issue is valid. However, according to our [Release Strategy](https://www.openssl.org/policies/releasestrat.html) the 1.0.2 branch is in security-fix-only mode, so this issue won't be fixed anymore.

Please note that the GOST engine has been removed in newer versions (see a8eda4312db1f98cffda38670e2d40d36566785a), but there is still a community version in an external repository. You might want to report the issue there: the error handling of both functions is inconsistent and both seem to leak the octet:

https://github.com/gost-engine/engine/blob/master/gost_ameth.c#L678-L683

https://github.com/gost-engine/engine/blob/master/gost_ameth.c#L753-L757


Regards,
Matthias

**Comment by mspncp** [comment]
cc @beldmit

**Comment by mspncp** [comment]
This issue won't be fixed in 1.0.2 anymore. But it has been reported in gost-engine/engine#125 and fixed by gost-engine/engine@2d267382074937ce48c150799102276a8842068f.
